### PR TITLE
templatize histogram

### PR DESCRIPTION
**Pull request by melvinw**

### CODE REVIEW

File: core/utils/histogram.h [hunk 1]
```diff
@@ -1,55 +1,145 @@
...
+      : num_buckets_(num_buckets),
+        bucket_width_(bucket_width),
+        threshold_(threshold),
+        buckets_(),
+        above_threshold_(),
+        count_(),
+        total_(),
+        min_bucket_(),
+        max_bucket_() {
+    buckets_ = static_cast<size_t *>(malloc(num_buckets_ * sizeof(size_t)));
```
Comment (barath) [inline core/utils/histogram.h]: Why not new[]?

File: core/utils/histogram.h [hunk 2]
```diff
@@ -1,55 +1,145 @@
...
-#define HISTO_TIME_TO_SEC(t) ((t) / (HISTO_TIMEUNIT_MULT / HISTO_TIME))
-#define PRINT_THRES (1)
-#define HISTO_BUCKET_VAL(ptr) ((*(ptr)))
-#define HISTO_BUCKET_INC(ptr) ((*(ptr))++)
+// A general purpose histogram. A bin b_i is labeled by its upper edge
+// T must be an arithmetic type
+template <typename T>
+class Histogram {
+ public:
+  Histogram(size_t num_buckets, T threshold, T bucket_width)
```
Comment (barath) [inline core/utils/histogram.h]: If it isn't too much work, could you add method comments for this and below.

Comment (melvinw) [inline core/utils/histogram.h]: Definitely. There's actually quite a bit of documentation i want to add. The interface exposed is not entirely obvious

File: core/utils/histogram.h (hunk 1 above)
Comment (melvinw) [inline core/utils/histogram.h]: Good catch. Will fix

File: core/utils/histogram.h [hunk 3]
```diff
@@ -1,55 +1,180 @@
...
+  size_t *buckets_;
+  size_t above_threshold_;
+  size_t count_;
+  T total_;
+  size_t min_bucket_;
+  size_t max_bucket_;
+};
+
+template <typename T>
+void Histogram<T>::summarize() {
```
Comment (barath) [inline core/utils/histogram.h]: Curious why you didn't just put this inline.

File: core/utils/histogram.h [hunk 4]
```diff
@@ -1,55 +1,180 @@
...
+      : num_buckets_(num_buckets),
+        bucket_width_(bucket_width),
+        threshold_((num_buckets_ + 1) * bucket_width_),
+        buckets_(),
+        above_threshold_(),
+        count_(),
+        total_(),
+        min_bucket_(),
+        max_bucket_() {
+    buckets_ = new size_t[num_buckets_];
```
Comment (barath) [inline core/utils/histogram.h]: Where does this get deleted?

File: core/utils/histogram.h (hunk 3 above)
Comment (melvinw) [inline core/utils/histogram.h]: It's a bit longer than the other methods. Looks weird all by itself on the outside, though. Will inline.

**Comment by sangjinhan** [comment]
For some reason, the 99%-ile (the third number) is higher than the average (the first number). It seems that the tail latency number is too low to be true...?

```
localhost:10514 $ run samples/latency
Tue Dec 27 15:25:46 2016: 6.829 Mpps, 3824.277 Mbps, 34.581 us, 30.000 us, 30.200 us
Tue Dec 27 15:25:47 2016: 6.724 Mpps, 3765.504 Mbps, 35.556 us, 29.900 us, 30.600 us
Tue Dec 27 15:25:48 2016: 6.561 Mpps, 3674.177 Mbps, 36.487 us, 30.100 us, 31.000 us
Tue Dec 27 15:25:49 2016: 6.642 Mpps, 3719.627 Mbps, 35.953 us, 30.700 us, 31.000 us
Tue Dec 27 15:25:50 2016: 6.611 Mpps, 3702.147 Mbps, 36.160 us, 30.600 us, 31.000 us
Tue Dec 27 15:25:51 2016: 6.669 Mpps, 3734.838 Mbps, 35.782 us, 29.700 us, 30.700 us
```

Also, it would be nice to see what each number represents for, like this:
```
Tue Dec 27 15:25:51 2016: 6.669 Mpps, 3734.838 Mbps, avg: 35.782 us, 50th: 29.700 us, 99th: 30.700 us
```

**Comment by sangjinhan** [comment]
(this issue happens with the non-template version too, so it is not a regression bug)

**Comment by melvinw** [comment]
pushed a fix. modules/measure was broken by 816dd07
```
localhost:10514 $ run samples/latency
Tue Dec 27 15:51:18 2016: 4.957 Mpps, 2776.049 Mbps, avg: 48.363 us, med: 43.500 us, 99th:106.100 us
Tue Dec 27 15:51:19 2016: 5.228 Mpps, 2927.744 Mbps, avg: 47.897 us, med: 43.500 us, 99th:96.100 us
Tue Dec 27 15:51:20 2016: 5.201 Mpps, 2912.670 Mbps, avg: 48.056 us, med: 44.100 us, 99th:98.600 us
Tue Dec 27 15:51:21 2016: 4.923 Mpps, 2756.869 Mbps, avg: 50.676 us, med: 47.100 us, 99th:100.400 us
Tue Dec 27 15:51:22 2016: 3.208 Mpps, 1796.592 Mbps, avg: 75.945 us, med: 47.600 us, 99th:182.200 us
Tue Dec 27 15:51:23 2016: 3.341 Mpps, 1870.720 Mbps, avg: 73.965 us, med: 48.800 us, 99th:186.200 us
Tue Dec 27 15:51:24 2016: 4.562 Mpps, 2554.533 Mbps, avg: 54.741 us, med: 47.900 us, 99th:130.000 us
```

**Comment by sangjinhan** [comment]
Merged, thanks!
6219047713_614df6a07d

**Pull request merged**